tests/testmplex.nim

```diff
@@ -594,6 +594,468 @@ suite "Mplex":
+      var mplexListen: Mplex
+      proc connHandler(conn: Connection) {.async, gcsafe.} =
+        mplexListen = Mplex.init(conn)
+        mplexListen.streamHandler = proc(stream: Connection)
+          {.async, gcsafe.} =
+          try:
+            discard await stream.readLp(1024)
+            check false
+          except LPStreamEOFError:
+            await stream.close()
+            return
+
+          check false
+
+        await mplexListen.handle()
+        await mplexListen.close()
+
+      let transport1 = TcpTransport.init()
+      let listenFut = await transport1.listen(ma, connHandler)
+
+      let transport2: TcpTransport = TcpTransport.init()
+      let conn = await transport2.dial(transport1.ma)
+
+      let mplexDial = Mplex.init(conn)
+      let mplexDialFut = mplexDial.handle()
+      var dialStreams: seq[Connection]
+      for i in 0..9:
+        dialStreams.add((await mplexDial.newStream()))
+
+      await mplexListen.close()
+
+      for s in dialStreams:
+        try:
+          discard await s.readLp(1024)
+          check false
+        except LPStreamEOFError:
+          check true
+          continue
+
+      await sleepAsync(1.seconds)
+      var channelTracker = getTracker(LPChannelTrackerName)
+      # echo channelTracker.dump()
+      check channelTracker.isLeaked() == false
+
+      await conn.close()
+      await mplexDialFut
+      await allFuturesThrowing(
+        transport1.close(),
+        transport2.close())
+      await listenFut
+
+    waitFor(testNewStream())
+
+  test "e2e - canceling mplex handler closes both ends":
+    proc testNewStream() {.async.} =
+      let ma: MultiAddress = Multiaddress.init("/ip4/0.0.0.0/tcp/0").tryGet()
+
+      var mplexHandle: Future[void]
+      proc connHandler(conn: Connection) {.async, gcsafe.} =
+        let mplexListen = Mplex.init(conn)
+        mplexListen.streamHandler = proc(stream: Connection)
+          {.async, gcsafe.} =
+          try:
+            discard await stream.readLp(1024)
+            check false
+          except LPStreamEOFError:
+            await stream.close()
+            return
+
+          check false
+
+        mplexHandle = mplexListen.handle()
+        await mplexHandle
+        await mplexListen.close()
+
+      let transport1 = TcpTransport.init()
+      let listenFut = await transport1.listen(ma, connHandler)
+
+      let transport2: TcpTransport = TcpTransport.init()
+      let conn = await transport2.dial(transport1.ma)
+
+      let mplexDial = Mplex.init(conn)
+      let mplexDialFut = mplexDial.handle()
+      var dialStreams: seq[Connection]
+      for i in 0..9:
+        dialStreams.add((await mplexDial.newStream()))
+
+      mplexHandle.cancel()
+
+      for s in dialStreams:
+        try:
+          discard await s.readLp(1024)
+          check false
+        except LPStreamEOFError:
+          check true
+          continue
+
+      await sleepAsync(1.seconds)
+      var channelTracker = getTracker(LPChannelTrackerName)
```

Comment:
this could be a template that prints the tracker on leak (or reports it to unittest) like the one in helpers